### PR TITLE
feat: Check Version before Extension upload

### PR DESCRIPTION
**Pull request by UnseenWizzard**
Add a step to extension_upload that looks for a currently deployed,
for the given extension name.

If the versions are equal, or the config as code version is lower,
skip the deployment, otherwise deploy the extension as before.

This removes the very user-unfriendly attempt to upload already
present versions, resulting in an incorrect "Error" log, and also
logs actual information about the already deployed version of the
extension.

fix: #125

### CODE REVIEW

File: pkg/rest/extension_upload_test.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		rw.Write([]byte(remotePayload))
+	}))
+	// Close the server when test finishes
+	defer server.Close()
+
+	status, err := validateIfExtensionShouldBeUploaded(server.Client(), server.URL, "name", []byte(localPayload), "token")
+	assert.Assert(t, err != nil)
+	assert.Equal(t, status, EXTENSION_VALIDATION_ERROR)
+}
```
Comment (patrickpichler) [inline pkg/rest/extension_upload_test.go:101]: A test with a missing remote version and local version would also be nice.

File: pkg/rest/extension_upload.go [hunk 2]
```diff
@@ -27,7 +29,26 @@ import (
...
 func uploadExtension(client *http.Client, apiPath string, extensionName string, payload []byte, apiToken string) (api.DynatraceEntity, error) {
+
+	status, err := validateIfExtensionShouldBeUploaded(client, apiPath, extensionName, payload, apiToken)
+	if err != nil {
+		return api.DynatraceEntity{}, err
+	}
+
+	if status == EXTENSION_UP_TO_DATE {
+		return api.DynatraceEntity{
+			Name: extensionName,
```
[the remark this answers is not in the document]
Comment (UnseenWizzard) [reply pkg/rest/extension_upload.go:48]: IIRC they had no accessible IDs before - on the new api that has changed, and with the introduced check we could actually get it. 
I'm not sure if the POST endpoint now also returns one though

Think this is something we could improve in a followup

File: pkg/rest/extension_upload.go [hunk 3]
```diff
@@ -27,7 +29,26 @@ import (
 	"github.com/dynatrace-oss/dynatrace-monitoring-as-code/pkg/util"
 )
 
+const (
+	EXTENSION_VALIDATION_ERROR = iota
+	EXTENSION_UP_TO_DATE       = iota
```
[the remark this answers is not in the document]
Comment (UnseenWizzard) [reply pkg/rest/extension_upload.go]: thx, fixed that

A type seems overkill to me though as this is just a status code "enum" returned by an internal method

[the remark this answers is not in the document]
Comment (patrickpichler) [reply pkg/rest/extension_upload.go]: Yes, but please define an enum as in:

```go
type ExtensionError int

const (
   ExtensionValidationError = iota
)
```

[the remark this answers is not in the document]
Comment (UnseenWizzard) [reply pkg/rest/extension_upload.go]: What kind of functional difference would an int type make to the method signature's typed return of `(status int, err error)`?

[the remark this answers is not in the document]
Comment (patrickpichler) [reply pkg/rest/extension_upload.go]: It would be more explicit/goesque.

**Review by patrickpichler — APPROVED** [review]

File: pkg/rest/extension_upload.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (UnseenWizzard) [reply pkg/rest/extension_upload.go]: just add the direct discussion here: 

I stand by this being overkill for an internal status code that does not leave the extension upload and it's test, but it's done

**Review by patrickpichler — APPROVED** [review]

**Pull request merged**